app/controllers/concerns/fully_authenticatable.rb

```diff
@@ -0,0 +1,14 @@
+module FullyAuthenticatable
+  def confirm_two_factor_authenticated(id = nil)
+    return redirect_to sign_up_start_url(request_id: id) unless user_signed_in?
+
+    return prompt_to_set_up_2fa unless current_user.two_factor_enabled?
+
+    prompt_to_enter_otp
```

Comment:
Probably not for this scenario, which is where a user makes a request from an SP while they are already signed in with email and password only and haven't entered their OTP yet. I'll add a test.